packages/model-viewer/src/features/scene-graph/model.ts

```diff
@@ -13,42 +13,275 @@
+    } else {
+      const mvMaterial = this[$materials].get(material);
+      if (mvMaterial != null) {
+        this.mesh.material = await mvMaterial[$ensureLoaded]();
+      }
+    }
+    return this.mesh.material;
+  }
+
+  async enableVariant(name: string):
+      Promise<ThreeMaterial|ThreeMaterial[]|null> {
+    if (this[$variantInfo] != null) {
+      const material = this[$variantInfo].get(name);
+      if (material != null) {
+        return await this.setActiveMaterial(material);
+      }
+    }
+    return null;
+  }
+
+  async instantiateVariants() {
+    if (this[$variantInfo] == null) {
+      return;
+    }
+    for (const name of this[$variantInfo].keys()) {
+      if (this.mesh.userData.variantMaterials.get(name).material != null) {
+        continue;
+      }
+      const threeMaterial = await this.enableVariant(name);
+      if (threeMaterial != null) {
+        this.mesh.userData.variantMaterials.get(name).material = threeMaterial;
+      }
+    }
+  }
+
+  get variantInfo() {
+    return this[$variantInfo];
+  }
+}
 
 /**
  * A Model facades the top-level GLTF object returned by Three.js' GLTFLoader.
  * Currently, the model only bothers itself with the materials in the Three.js
  * scene graph.
  */
 export class Model implements ModelInterface {
-  private[$materials]: Array<Material> = [];
+  private[$materials] = new Array<Material>();
+  private[$hierarchy] = new Array<Node>();
+  private[$roots] = new Array<Node>();
+  private[$primitives] = new Array<PrimitiveNode>();
 
   constructor(
       correlatedSceneGraph: CorrelatedSceneGraph,
       onUpdate: () => void = () => {}) {
-    const {gltf, gltfElementMap} = correlatedSceneGraph;
+    const {gltf, threeGLTF, gltfElementMap} = correlatedSceneGraph;
+
+    for (let i = 0, size = gltf.materials!.length; i < size; ++i) {
+      const material = gltf.materials![i];
 
-    gltf.materials!.forEach(material => {
       const correlatedMaterial =
           gltfElementMap.get(material) as Set<MeshStandardMaterial>;
 
       if (correlatedMaterial != null) {
         this[$materials].push(
             new Material(onUpdate, gltf, material, correlatedMaterial));
       } else {
-        console.warn(`Unreferenced material "${
-            material.name}" was automatically removed from the model.`);
+        const elementArray = gltf['materials'] || [];
+        const gltfMaterialDef = elementArray[i];
+
+        // Loads the three.js material.
+        const capturedMatIndex = i;
+        const materialLoadCallback = async () => {
+          const threeMaterial =
+              await threeGLTF.parser.getDependency(
+                  'material', capturedMatIndex) as MeshStandardMaterial;
+
+          // Adds correlation, maps the variant gltf-def to the
+          // three material set containing the variant material.
+          const threeMaterialSet = new Set<MeshStandardMaterial>();
+          gltfElementMap.set(gltfMaterialDef, threeMaterialSet);
```

Comment:
Do we know that every invocation of this has a unique gltfMaterialDef? I just want to make sure we're safe always creating a new set rather than adding to an existing one.

Reply:
Yes, the only way it ends up here is if that was missing in the first place.